streams/src/main/java/org/apache/kafka/streams/state/internals/ReadonlyPartiallyDeserializedSegmentValue.java

```diff
@@ -0,0 +1,210 @@
+
+    private static final int TIMESTAMP_SIZE = 8;
+    private static final int VALUE_SIZE = 4;
+    private byte[] segmentValue;
+    private long nextTimestamp;
+    private long minTimestamp;
+
+    private int deserIndex = -1; // index up through which this segment has been deserialized (inclusive)
+
+    private Map<Integer, Integer> cumulativeValueSizes;
+
+    private int valuesStartingIndex = -1; // the index of the first value in the segment (but the last one in the list)
+    private Map<Integer, TimestampAndValueSize> unpackedTimestampAndValueSizes = new HashMap<>();
+    private int recordNumber = -1; // number of segment records
+
+
+    ReadonlyPartiallyDeserializedSegmentValue(final byte[] segmentValue) {
+        this.segmentValue = segmentValue;
+        this.nextTimestamp =
+                RocksDBVersionedStoreSegmentValueFormatter.getNextTimestamp(segmentValue);
+        this.minTimestamp =
+                RocksDBVersionedStoreSegmentValueFormatter.getMinTimestamp(segmentValue);
+        resetDeserHelpers();
+    }
+
+
+    public long minTimestamp() {
+        return minTimestamp;
+    }
+
+    public long nextTimestamp() {
+        return nextTimestamp;
+    }
+
+    public byte[] serialize() {
+        return segmentValue;
+    }
+
+
+    public RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult find(
+            final long fromTime, final long toTime, final ResultOrder order, final int index) {
+
+        // this segment does not have any record in query specified time range
+        if (toTime < minTimestamp || fromTime > nextTimestamp) {
+            return null;
+        }
+
+        final boolean isAscending = order.equals(ResultOrder.ASCENDING);
+
+        if (isAscending && valuesStartingIndex == -1) {
+            findValuesStartingIndex();
+            deserIndex = recordNumber;
+        }
+
+        long currTimestamp = -1;
+        long currNextTimestamp = -1;
+        int currIndex = initializeCurrentIndex(index, isAscending);
+        int cumValueSize = initializeCumValueSize(index, currIndex, isAscending);
+        int currValueSize;
+
+
+        while (hasStillRecord(currTimestamp, currNextTimestamp, order)) {
+            if (hasBeenDeserialized(isAscending, currIndex)) {
+                final TimestampAndValueSize curr;
+                curr = unpackedTimestampAndValueSizes.get(currIndex);
+                currTimestamp = curr.timestamp;
+                cumValueSize = cumulativeValueSizes.get(currIndex);
+                currValueSize = curr.valueSize;
+
+                // update currValueSize
+                if (currValueSize == Integer.MIN_VALUE) {
+                    final int timestampSegmentIndex = getTimestampIndex(order, currIndex);
+                    currValueSize = ByteBuffer.wrap(segmentValue).getInt(timestampSegmentIndex + TIMESTAMP_SIZE);
+                    unpackedTimestampAndValueSizes.put(currIndex, new TimestampAndValueSize(currTimestamp, cumValueSize));
```

Comment:
`cumValueSize` is not updated with `currValueSize`. Why?

Reply:
> `cumValueSize` is not updated with `currValueSize`. Why?

Yes, in fact in this if (if deserIndex is ahead of currIndex), we must have EVERYTHING ready. Therefore, `if (currValueSize == Integer.MIN_VALUE)` does not fit here. I removed it.